### PR TITLE
Compiler Cleanup and Metadata Helper Functions

### PR DESCRIPTION
**Pull request by InsertCreativityHere**
This PR removes repetitive logic from the compilers:
- Moved our `SerialVersionUUID` logic into a function for slice2java (instead of having the same thing copied 3 times).
- Moved the 3 `getDeprecateReason` functions scattered around the compilers into a single version in SliceUtil (they were nearly identical anyways).
- Removed the `checkFileMetadata` method, which isn't used anywhere.

It also adds the following helper methods into SliceUtil:
- `parseMetadata`
- `findMetadata`
- `hasMetadata`
These are only needed by the new `getDeprecateReason` currently. In my 'real' PR they're all used extensively. Feel free to critique their implementation though.
I'm opening this PR first to keep the 'cleanup' somewhat separated from the rest of my changes.

### CODE REVIEW

File: cpp/src/Slice/SliceUtil.cpp [hunk 1]
```diff
@@ -788,3 +788,72 @@ bool Slice::opCompressReturn(const OperationPtr& op)
...
+    }
+    else
+    {
+        // Make sure the parentheses are balanced. For metadata with arguments, the final character must be ')'.
+        auto argumentEnd = metadata.length() - 1;
+        if (metadata.at(argumentEnd) != ')')
+        {
+            unit->error("missing closing parentheses for metadata arguments");
+        }
+        return make_pair(metadata.substr(0, argumentStart), metadata.substr(argumentStart + 1, argumentEnd));
```
Comment (pepone) [inline cpp/src/Slice/SliceUtil.cpp:832]: `argumentEnd` doesn't look correct here, we are counting from the beginning but should be counting from `argumentStart +1` 

What about something like:
```
auto start = metadata.find("(");
if (start == string::npos)
{
    return make_pair(metadata, "");
}
else if (!metadata.ends_with(')'))
{
    unit->error("missing closing parentheses for metadata arguments");
    return make_pair(metadata.substr(0, start), "");
}
else
{
    return make_pair(metadata.substr(0, start),
                     string(next(metadata.begin() + start), prev(metadata.end())));
}
```

File: cpp/src/Slice/SliceUtil.cpp [hunk 2]
```diff
@@ -788,3 +788,72 @@ bool Slice::opCompressReturn(const OperationPtr& op)
...
+}
+
+optional<string> Slice::findMetadata(const string& directive, const map<string, string>& metadata)
+{
+    auto match = metadata.find(directive);
+    if (match != metadata.end())
+    {
+        return match->second;
+    }
+    return nullopt;
```
Comment (pepone) [inline cpp/src/Slice/SliceUtil.cpp:858]: Can we use the ternary operator for these simple conditions?

```suggestion
    return match != metadata.end() ? match->second : nullopt;
```

File: cpp/src/Slice/SliceUtil.cpp [hunk 3]
```diff
@@ -788,3 +788,72 @@ bool Slice::opCompressReturn(const OperationPtr& op)
 {
     return opCompress(op, false);
 }
+
+string Slice::getDeprecateReason(const ContainedPtr& p, bool checkContainer)
```
Comment (pepone) [inline cpp/src/Slice/SliceUtil.cpp:792]: I wondering if we should use an scoped enum `CheckContainter {yes, no}` to make code more readable

```
getDeprecateReason(p, CheckContainer::no);

getDeprecateReason(p, false);
```

**Review by pepone — APPROVED** [review]

File: cpp/src/Slice/SliceUtil.cpp [hunk 4]
```diff
@@ -788,3 +788,72 @@ bool Slice::opCompressReturn(const OperationPtr& op)
...
+{
+    // Check if there was `deprecate` metadata directly on the entity.
+    auto deprecateMsg = p->findMetadata("deprecate");
```
[the remark this answers is not in the document]
Comment (InsertCreativityHere) [reply cpp/src/Slice/SliceUtil.cpp:795]: Fixed!

File: cpp/src/Slice/SliceUtil.cpp [hunk 5]
```diff
@@ -788,3 +788,72 @@ bool Slice::opCompressReturn(const OperationPtr& op)
...
+    {
+        return make_pair(metadata, "");
+    }
+    else
+    {
+        // Make sure the parentheses are balanced. For metadata with arguments, the final character must be ')'.
+        auto argumentEnd = metadata.length() - 1;
+        if (metadata.at(argumentEnd) != ')')
+        {
+            unit->error("missing closing parentheses for metadata arguments");
```
[the remark this answers is not in the document]
Comment (InsertCreativityHere) [reply cpp/src/Slice/SliceUtil.cpp:830]: I think it's better to return the actual metadata. We know what the metadata is, the user just made a small syntactical mistake. Plus, by returning the actual metadata this gives us the chance to check for invalid metadata, and I think the sooner we can give the user errors, the sooner they can fix them.

File: cpp/src/Slice/SliceUtil.cpp (hunk 2 above)
Comment (InsertCreativityHere) [reply cpp/src/Slice/SliceUtil.cpp:858]: Yes, absolutely!

File: cpp/src/Slice/SliceUtil.cpp (hunk 3 above)
Comment (InsertCreativityHere) [reply cpp/src/Slice/SliceUtil.cpp:792]: To me personally, introducing a new enum just for this seems a little overkill. I think it's easy enough to just find the source (or header once we name the parameters in them) and see what the name is.

In an ideal world, we could use named parameters, but C++ doesn't natively support them. Maybe this is also overkill, but there are ways of emulating named parameters by overloading the '=' operator to get stuff like this:
```
using CheckContainer = NamedWrapperThing<bool>;
...
getDeprecateReason(p, CheckContainer = false);
```

**Pull request merged**